File: TcpSocketSystem_Java/src/com/tecnm/sockets/infrastructure/network/TcpClienteSocket.java
package com.tecnm.sockets.infrastructure.network;

import com.tecnm.sockets.domain.puertos.IServicioLogging;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

/**
 * Cliente de Sockets TCP en Java con política de reintentos, timeouts configurables y cierre limpio.
 */
public final class TcpClienteSocket implements AutoCloseable {

    private final IServicioLogging logger;
    private Socket socket;
    private BufferedReader reader;
    private BufferedWriter writer;
    private volatile boolean cerrado;

    public TcpClienteSocket(IServicioLogging logger) {
        if (logger == null) {
            throw new IllegalArgumentException("El logger no puede ser nulo.");
        }
        this.logger = logger;
        this.cerrado = false;
    }

    public boolean conectarConReintentos(String host, int puerto, int maxReintentos, int delayMs) {
        for (int intento = 1; intento <= maxReintentos; intento++) {
            try {
                logger.info(String.format("Intentando conectar a %s:%d (Intento %d/%d)...", host, puerto, intento, maxReintentos), "TcpClienteSocket");

                this.socket = new Socket();
                this.socket.setTcpNoDelay(true);
                this.socket.setSoTimeout(10000); // 10s timeout de lectura
                this.socket.connect(new InetSocketAddress(host, puerto), 5000); // 5s timeout de conexión

                this.reader = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8));
                this.writer = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream(), StandardCharsets.UTF_8));
                this.cerrado = false;

                logger.info(String.format("¡Conectado exitosamente al servidor %s:%d!", host, puerto), "TcpClienteSocket");
                return true;

            } catch (IOException ex) {
                logger.warning(String.format("Fallo en intento %d: %s", intento, ex.getMessage()), "TcpClienteSocket");
                desconectar();

                if (intento < maxReintentos) {
                    try {
                        Thread.sleep(delayMs);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        break;
                    }
                }
            }
        }

        logger.error(String.format("No se pudo conectar a %s:%d tras %d intentos.", host, puerto, maxReintentos), null, "TcpClienteSocket");
        return false;
    }

    public boolean estaConectado() {
        return socket != null && socket.isConnected() && !socket.isClosed() && !cerrado;
    }

    public String enviarYRecibirEco(String mensaje) {
        if (!estaConectado()) {
            throw new IllegalStateException("El cliente no está conectado.");
        }

        try {
            logger.info(String.format("[ENVIANDO] -> \"%s\"", mensaje), "TcpClienteSocket");
            writer.write(mensaje);
            writer.write("\n");
            writer.flush();

            String respuesta = reader.readLine();
            logger.info(String.format("[RECIBIDO] <- \"%s\"", respuesta), "TcpClienteSocket");
            return respuesta;
        } catch (IOException ex) {
            logger.error("Error de comunicación durante el envío/recepción.", ex, "TcpClienteSocket");
            desconectar();
            return null;
        }
    }

    public void desconectar() {
        if (cerrado) return;
        cerrado = true;

        try { if (writer != null) writer.close(); } catch (Exception ignored) {}
        try { if (reader != null) reader.close(); } catch (Exception ignored) {}
        try { if (socket != null) socket.close(); } catch (Exception ignored) {}

        this.writer = null;
        this.reader = null;
        this.socket = null;
    }

    @Override
    public void close() {
        desconectar();
    }
}

File: TcpSocketSystem_DotNet/src/Infrastructure/Network/TcpClienteSocket.cs
using System.Net.Sockets;
using System.Text;
using TcpSocketSystem.Core.Domain.Puertos;

namespace TcpSocketSystem.Core.Infrastructure.Network;

/// <summary>
/// Cliente TCP con políticas de reintento, timeout configurable, sanitización y cierre limpio.
/// </summary>
public sealed class TcpClienteSocket : IDisposable
{
    private readonly IServicioLogging _logger;
    private TcpClient? _cliente;
    private NetworkStream? _stream;
    private StreamReader? _reader;
    private StreamWriter? _writer;
    private bool _disposed;

    public bool EstaConectado => _cliente?.Connected == true && !_disposed;

    public TcpClienteSocket(IServicioLogging logger)
    {
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
    }

    public async Task<bool> ConectarConReintentosAsync(string host, int puerto, int maxReintentos = 3, int delayMs = 1500, CancellationToken cancellationToken = default)
    {
        for (int intento = 1; intento <= maxReintentos; intento++)
        {
            try
            {
                _logger.Info($"Intentando conectar a {host}:{puerto} (Intento {intento}/{maxReintentos})...", nameof(TcpClienteSocket));
                
                _cliente = new TcpClient
                {
                    NoDelay = true,
                    ReceiveTimeout = 10000,
                    SendTimeout = 10000
                };

                using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
                cts.CancelAfter(TimeSpan.FromSeconds(5));

                await _cliente.ConnectAsync(host, puerto, cts.Token);
                
                _stream = _cliente.GetStream();
                _reader = new StreamReader(_stream, new UTF8Encoding(false), leaveOpen: false);
                _writer = new StreamWriter(_stream, new UTF8Encoding(false))
                {
                    AutoFlush = true,
                    NewLine = "\n"
                };

                _logger.Info($"¡Conectado exitosamente al servidor {host}:{puerto}!", nameof(TcpClienteSocket));
                return true;
            }
            catch (Exception ex)
            {
                _logger.Warning($"Fallo al conectar en el intento {intento}: {ex.Message}", nameof(TcpClienteSocket));
                Desconectar();

                if (intento < maxReintentos)
                {
                    await Task.Delay(delayMs, cancellationToken);
                }
            }
        }

        _logger.Error($"No fue posible establecer conexión con {host}:{puerto} tras {maxReintentos} intentos.", origen: nameof(TcpClienteSocket));
        return false;
    }

    public async Task<string?> EnviarYRecibirEcoAsync(string mensaje, CancellationToken cancellationToken = default)
    {
        if (!EstaConectado || _writer == null || _reader == null)
        {
            throw new InvalidOperationException("El cliente no está conectado a ningún servidor.");
        }

        try
        {
            _logger.Info($"[ENVIANDO] -> \"{mensaje}\"", nameof(TcpClienteSocket));
            await _writer.WriteLineAsync(mensaje.AsMemory(), cancellationToken);

            var respuesta = await _reader.ReadLineAsync(cancellationToken);
            _logger.Info($"[RECIBIDO] <- \"{respuesta}\"", nameof(TcpClienteSocket));
            return respuesta;
        }
        catch (IOException ex)
        {
            _logger.Error("Conexión interrumpida durante la transmisión de datos.", ex, nameof(TcpClienteSocket));
            Desconectar();
            return null;
        }
    }

    public void Desconectar()
    {
        try { _writer?.Dispose(); } catch { }
        try { _reader?.Dispose(); } catch { }
        try { _stream?.Dispose(); } catch { }
        try { _cliente?.Close(); } catch { }
        try { _cliente?.Dispose(); } catch { }

        _writer = null;
        _reader = null;
        _stream = null;
        _cliente = null;
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        Desconectar();
    }
}
